README.md
# AdjeminPay Flutter

With AdjeminPay, integrating mobile money payments in your flutter app becomes super easy, barely an inconvenience.

## Using

The package generates a payment gate, and when the pay user is finished paying, the payment gate return the payment result.

Follow the steps below :

### Step 1. Get and store your credentials in a constants files

Sign up and create a merchant application for free at [AdjeminPay][] to get an apiKey and an applicationId.

[AdjeminPay]: https://merchant.adjeminpay.net

Then store them in a config/constants/env file where you keep all your other constant data.

````dart
class Constants{

  static final String ADJEMINPAY_CLIENT_ID = "CLIENT_ID";
  static final String ADJEMINPAY_CLIENT_SECRET = "CLIENT_SECRET";
  static final String ADJEMINPAY_NOTIFICATION_URL = "https://your.backend.api/notifyUrl";
}
````

### Step 2. Import the package

Add the adjeminpay_flutter package as a dependancy in your `pubspec.yaml` file.

```yaml
    dependencies:
    flutter:
        sdk: flutter
    adjeminpay_flutter: ^1.1.103
```

Then import it to your cart screen

```dart
import 'package:adjeminpay_flutter/adjeminpay_flutter.dart';

```

Your `cart_screen` is where you will show your users the items they will pay for.

### Step 3. Build your cart screen

Build a cart screen to finalize the payment.

You can find a beautiful cart screen example from [Academind][] by Maximilian Schwarzmüller in the [Example][] file.

[Academind]: https://academind.com/
[Example]: https://https://github.com/adjemin/adjeminpay_flutter/tree/master/example

### Step 4. Have your order/transaction data ready

Store the data about the payment your user is about to make.

NB: _It is recommended you store the payment data in your database before executing the payment function._

This example uses a basic Map for storing data about an order but it could be anything, from transactions, to payment, anything that needs online payment.

```dart
    Map<String, dynamic> myOrder = {
      // ! required transactionId or orderId
      'transaction_id': "UniqueTransactionId",
      // ! required total amount
      'total_amount': 1000,
      // optional your orderItems data
      'items': [
        {
          'id': '1',
          'productId': 'prod1',
          'price': 100,
          'quantity': 1,
          'title': 'Product 1 title',
        },
        {
          'id': '2',
          'productId': 'prod9',
          'price': 300,
          'quantity': 3,
          'title': 'Product 9 title',
        },
      ],
      'currency_code': "XOF",
      'designation': "Order Title",
      'client_name': "ClientName",
    };
```

### Step 5. Add a function to launch the payment gate and handle the payment result

Copy-paste this function that takes in the order data and extracts the fields needed by the payment api.

`required fields`:
The following fiels are required:

* `clientId`      : You can get it in `Step 1.`
* `clientSecret`    : You can get it in `Step 1.`
* `merchantTransactionId` : Unique id for everyone of your payments
* `designation` : What your user will see as what they're paying for
* `amount`  : Obviously, right ?
* `currencyCode`: Currency for the payment
  NB: _As of version ^0.1 only `XOF` is supported, so check back soon for more_
* `notificationUrl` : A url you set in your web backend.
  A post request with the payment result (a json containing {transactionId, status, message}) will be sent to that url when the payment is completed (successful, failed, cancelled, expired) to allow you to update the order/transaction status in your database directly from your backend.
  This field is optional, since you could do such an update as a callback from your flutter app.
  NB: _If you do want to use the notificationUrl, please check out our [php sdk](https://github.com/adjemin/adjeminpay-php-sdk) for how to catch the notification._
 
`optional fields`:
These fields are optional:
* `buyerName` : The name of your user
  NB: _It is recommended that you provide it in your payment data as it saves your user the hassle of having to type it in the payment gate._

Now let's implement the payment handler function

NB: _This function and the callbacks will be included directly in the library in future versions._

```dart
void payWithAdjeminPay(dynamic orderData) async {
      // paymentResult will yield {transactionId, status, message }
      // once the payment gate is closed by the user
      // ! IMPORTANT : make sure to save the orderData in your database first
      // ! before calling this function
      Map<String, dynamic> paymentResult = await Navigator.push(
          context,
          new MaterialPageRoute(
            builder: (context) =>
                // The AdjeminPay class
                AdjeminPay(
                   // ! required clientId
                  clientId: Constants.ADJEMINPAY_CLIENT_ID,
                  // ! required clientScret
                  clientSecret: Constants.ADJEMINPAY_CLIENT_SECRET,
                  // ! required transactionId required
                  // for you to follow the transaction
                  //    or retrieve it later
                  //    should be a string < 191 and unique for your application
                  merchantTransactionId: "${orderData['transaction_id']}",
                      // ! required designation
                      //   the name the user will see as what they're paying for
                  designation: orderData['designation'],
                  // notifyUrl for your web backend
                  notificationUrl:Constants.ADJEMINPAY_NOTIFICATION_URL,
                  // amount: int.parse("${orderData['totalAmount']}"),
                  // ! required amount
                  //    amount the user is going to pay
                  //    should be an int
                  amount: int.parse("${orderData['total_amount']}"),
                  // currency code
                  // currently supported currency is XOF
                  currencyCode: orderData['currency_code'],
                  // designation: widget.element.title,
                  // the name of your user
                  buyerName: orderData['client_name'], //optional  
                ),
          ));

      print(">>> ADJEMINPAY PAYMENT RESULTS <<<");
      print(paymentResult);
      // * Here you define your callbacks
      // Callback if the paymentResult is null
      //    the payment gate got closed without sending back any data
      if (paymentResult == null) {
        print("<<< Payment Gate Unexpectedly closed");
        return;
      }
      Scaffold.of(context).showSnackBar(SnackBar(content: Text("Payment Status is ${paymentResult['status']}")));
      // Callback on payment successfully
       if (paymentResult['status'] == "SUCCESSFUL") {
        print("<<< AdjeminPay success");
        print(paymentResult);
        // redirect to or show another screen
        return;
      }
      // Callback on payment failed
      if (paymentResult['status'] == "FAILED") {
        print("<<< AdjeminPay failed");
        print(paymentResult);
        // the reason with be mentionned in the paymentResult['message']
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on payment cancelled
      if (paymentResult['status'] == "CANCELLED") {
        print("<<< AdjeminPay cancelled");
        print(paymentResult);
        // the reason with be mentionned in the paymentResult['message']
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on payment cancelled
      if (paymentResult['status'] == "EXPIRED") {
        print("<<< AdjeminPay expired");
        print(paymentResult);
        // The user took too long to approve or refuse payment
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on initialisation error
      if (paymentResult['status'] == "INVALID_PARAMS") {
        print("<<< AdjeminPay Init error");
        // You didn't specify a required field
        print(paymentResult);
        return;
      }

      if (paymentResult['status'] == "INVALID_CREDENTIALS") {
        print("<<< AdjeminPay Init error");
        // You didn't specify a required field
        // or your clientId or clientSecret are not valid
        print(paymentResult);
        return;
      }
      // Callback when AdjeminPay requests aren't completed
      if (paymentResult['status'] == "ERROR_HTTP") {
        return;
      }
    }
```


### Step 6. Pass the function above to your "Order Now" button

```dart
FlatButton(
child: _isLoading
    ? CircularProgressIndicator()
    : Text('ORDER NOW'),
onPressed: (myOrder['totalAmount'] == null ||
        myOrder['totalAmount'] <= 0 ||
        _isLoading)
    ? null
    : () {
        // **** Payment Management Here
        // you first store the Order's data in
        // your database where you create a unique transaction Id
        // for example :  await storeOrderData(myOrder);
        // then you call the payment function
        payWithAdjeminPay(myOrder);
        },
textColor: Theme.of(context).primaryColor,
);
```

### Step 7. You're all done

Congratulations ! You just integrated mobile money payment to your flutter app. Happy coding \(><)/ !

### Complete Example

You can find a basic example at [Example][]

Please help us improve and file an [issue](https://github.com/adjemin/adjeminpay_flutter/issues)


CHANGELOG.md
# ChangeLog AdjeminPay Flutter

## [1.1.103] - 14/06/2021
* Bug fixed

## [1.1.102] - 14/06/2021
* Bug fixed

## [0.1.7] - 12/06/2021
* AdjeminPay APis V2  integrated
* Bug fixed

## [0.1.6] - 01/30/2021

* 10 digits validation added

## [0.1.5] - 01/30/2021

* Added PNN support for Ivorian Numbers

## [0.1.4] - 01/05/2021

* Transaction close from pending screen
* Improved user experience for mtn-ci status update unusual duration

## [0.1.3] - 11/17/2020

* Added manual transaction status refresh button

## [0.1.2+1] - 11/17/2020

* Code refractor for static analysis.

## [0.1.2] - 11/09/2020

* Added payment abort feature.
  
## [0.0.1] - 09/17/2020

* Basic Payment Gate generation from credentials and payment order data.


example/example.dart
import 'package:flutter/material.dart';
import '../lib/adjeminpay_flutter.dart';
import 'dart:math' as m;

// This page describes a very basic use of AdjeminPay
// to finalize an order on a cart screen

class ExampleScreen extends StatefulWidget {
  // The route name
  static const routeName = '/adjeminpay-example';

  @override
  _ExampleScreenState createState() => _ExampleScreenState();
}

class _ExampleScreenState extends State<ExampleScreen> {
  // To either display a loading spinner or the "Order Now" button
  var _isLoading = false;

  static final String clientId = "CLIENT_ID";
  static final String clientSecret = "CLIENT_SECRET";

  static String generateUUID() {
    var rnd = new m.Random.secure();

    var bytes = new List<int>.generate(16, (_) => rnd.nextInt(256));
    bytes[6] = (bytes[6] & 0x0F) | 0x40;
    bytes[8] = (bytes[8] & 0x3f) | 0x80;

    var chars = bytes
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join()
        .toUpperCase();

    return '${chars.substring(0, 8)}-${chars.substring(8, 12)}-'
        '${chars.substring(12, 16)}-${chars.substring(16, 20)}-${chars.substring(20, 32)}';
  }

  @override
  Widget build(BuildContext context) {
    // The basic configuration of AdjeminPay
    // You need to keep your apiKey and applicationId secret
    // You can store them in a constants.dart or env.dart file
    // They are required for all the operations with the package
    // This functions passes in data (your order/transaction data)
    // to the AdjeminPay() constructor, that will generate a payment gateway
    // where your user will enter their :
    // - name (it is recommended that
    //        you pass the name as an argument to AdjeminPay() to
    //        save your user the incovenience of having to type it in ),
    // - mobile money phone number
    // - or credit card information
    //
    // The payment gateway will then:
    // - notify your notifyUrl (if provided)
    //        with the transactionId, status and message
    // - return a Map containing transactionId, status and message
    //        for you to execute any callback or redirection accordingly

    void payWithAdjeminPay(dynamic orderData, String clientId, String clientSecret) async {
      // paymentResult will yield {transactionId, status, message }
      // once the payment gate is closed by the user
      // ! IMPORTANT : make sure to save the orderData in your database first
      // ! before calling this function
      Map<String, dynamic> paymentResult = await Navigator.push(
          context,
          new MaterialPageRoute(
            builder: (context) =>
                // The AdjeminPay class
                AdjeminPay(
              // ! required clientId
              clientId: clientId,
              // ! required clientScret
              clientSecret: clientSecret,
              // ! required transactionId required
              // for you to follow the transaction
              //    or retrieve it later
              //    should be a string and unique for your application
              merchantTransactionId: "${orderData['transaction_id']}",
                  // ! required designation
                  //   the name the user will see as what they're paying for
              designation: orderData['designation'],
              // notifyUrl for your web backend
              notificationUrl: "https://api.exemple.com/epayment/callback",
              // amount: int.parse("${orderData['totalAmount']}"),
              // ! required amount
              //    amount the user is going to pay
              //    should be an int
              amount: int.parse("${orderData['total_amount']}"),
              // currency code
              // currently supported currency is XOF
              currencyCode: orderData['currency_code'],
              // designation: widget.element.title,
              // the name of your user
              buyerName: orderData['client_name'], //optional
            ),
          ));

      print(">>> ADJEMINPAY PAYMENT RESULTS <<<");
      print(paymentResult);
      // * Here you define your callbacks
      // Callback if the paymentResult is null
      //    the payment gate got closed without sending back any data
      if (paymentResult == null) {
        print("<<< Payment Gate Unexpectedly closed");
        return;
      }
      Scaffold.of(context).showSnackBar(SnackBar(content: Text("Payment Status is ${paymentResult['status']}")));
      // Callback on payment successfully
      if (paymentResult['status'] == "SUCCESSFUL") {
        print("<<< AdjeminPay success");
        print(paymentResult);
        // redirect to or show another screen
        return;
      }
      // Callback on payment failed
      if (paymentResult['status'] == "FAILED") {
        print("<<< AdjeminPay failed");
        print(paymentResult);
        // the reason with be mentionned in the paymentResult['message']
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on payment cancelled
      if (paymentResult['status'] == "CANCELLED") {
        print("<<< AdjeminPay cancelled");
        print(paymentResult);
        // the reason with be mentionned in the paymentResult['message']
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on payment cancelled
      if (paymentResult['status'] == "EXPIRED") {
        print("<<< AdjeminPay expired");
        print(paymentResult);
        // The user took too long to approve or refuse payment
        print("Reason is : " + paymentResult['message']);
        // redirect to or show another screen
        return;
      }
      // Callback on initialisation error
      if (paymentResult['status'] == "INVALID_PARAMS") {
        print("<<< AdjeminPay Init error");
        // You didn't specify a required field
        print(paymentResult);
        return;
      }

      if (paymentResult['status'] == "INVALID_CREDENTIALS") {
        print("<<< AdjeminPay Init error");
        // You didn't specify a required field
        // or your clientId or clientSecret are not valid
        print(paymentResult);
        return;
      }
      // Callback when AdjeminPay requests aren't completed
      if (paymentResult['status'] == "ERROR_HTTP") {
        return;
      }
    }

    // Map of a basic order data,
    // recommanded to use your Order/Cart class
    Map<String, dynamic> myOrder = {
      // ! required transactionId or orderId
      'transaction_id': generateUUID().toLowerCase(),
      // ! required total amount
      'total_amount': 1000,
      // optional your orderItems data
      'items': [
        {
          'id': '1',
          'product_id': 'prod1',
          'price': 100,
          'quantity': 1,
          'title': 'Product 1 title',
        },
        {
          'id': '2',
          'product_id': 'prod9',
          'price': 300,
          'quantity': 3,
          'title': 'Product 9 title',
        },
      ],
      'currency_code': "XOF",
      'designation': "Order Title",
      'client_name': "Ange Bagui",
      'client_phone':"2250556888385"
    };

    // Basic Cart Screen for checkout from Academind by Maximilian Schwarzmüller
    // https://academind.com/

    return Scaffold(
      appBar: AppBar(
        title: Text('Your Cart'),
      ),
      body: Column(
        children: <Widget>[
          Card(
            margin: EdgeInsets.all(15),
            child: Padding(
              padding: EdgeInsets.all(8),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  Text(
                    'Total',
                    style: TextStyle(fontSize: 20),
                  ),
                  Spacer(),
                  Chip(
                    label: Text(
                      '\$${myOrder['total_amount']?.toStringAsFixed(2)}',
                      style: TextStyle(
                        color: Theme.of(context).primaryTextTheme.title.color,
                      ),
                    ),
                    backgroundColor: Theme.of(context).primaryColor,
                  ),
                  FlatButton(
                    child: _isLoading
                        ? CircularProgressIndicator()
                        : Text('ORDER NOW'),
                    onPressed: (myOrder['total_amount'] == null ||
                            myOrder['total_amount'] <= 0 ||
                            _isLoading)
                        ? null
                        : () {
                            // **** Payment Management Here
                            // you first store the Order's data in
                            // your database where you create a unique transaction Id
                            // for example :  await storeOrderData(myOrder);
                            // then you call the payment function
                            payWithAdjeminPay(myOrder, clientId,clientSecret);
                          },
                    textColor: Theme.of(context).primaryColor,
                  )
                ],
              ),
            ),
          ),
          SizedBox(height: 10),
          Expanded(
            child: ListView.builder(
              itemCount: myOrder['items'].length,
              itemBuilder: (ctx, i) => CartItem(
                myOrder['items'][i]['id'],
                myOrder['items'][i]['product_id'],
                myOrder['items'][i]['price'],
                myOrder['items'][i]['quantity'],
                myOrder['items'][i]['title'],
              ),
            ),
          )
        ],
      ),
    );
  }
}

class CartItem extends StatelessWidget {
  final String id;
  final String productId;
  final int price;
  final int quantity;
  final String title;

  CartItem(
    this.id,
    this.productId,
    this.price,
    this.quantity,
    this.title,
  );

  @override
  Widget build(BuildContext context) {
    return Dismissible(
      key: ValueKey(id),
      background: Container(
        color: Theme.of(context).errorColor,
        child: Icon(
          Icons.delete,
          color: Colors.white,
          size: 40,
        ),
        alignment: Alignment.centerRight,
        padding: EdgeInsets.only(right: 20),
        margin: EdgeInsets.symmetric(
          horizontal: 15,
          vertical: 4,
        ),
      ),
      direction: DismissDirection.endToStart,
      confirmDismiss: (direction) {
        return showDialog(
          context: context,
          builder: (ctx) => AlertDialog(
            title: Text('Are you sure?'),
            content: Text(
              'Do you want to remove the item from the cart?',
            ),
            actions: <Widget>[
              FlatButton(
                child: Text('No'),
                onPressed: () {
                  Navigator.of(ctx).pop(false);
                },
              ),
              FlatButton(
                child: Text('Yes'),
                onPressed: () {
                  Navigator.of(ctx).pop(true);
                },
              ),
            ],
          ),
        );
      },
      onDismissed: (direction) {
        //
      },
      child: Card(
        margin: EdgeInsets.symmetric(
          horizontal: 15,
          vertical: 4,
        ),
        child: Padding(
          padding: EdgeInsets.all(8),
          child: ListTile(
            leading: CircleAvatar(
              child: Padding(
                padding: EdgeInsets.all(5),
                child: FittedBox(
                  child: Text('\$$price'),
                ),
              ),
            ),
            title: Text(title),
            subtitle: Text('Total: \$${(price * quantity)}'),
            trailing: Text('$quantity x'),
          ),
        ),
      ),
    );
  }
}
